### PR TITLE
Fetch messages

### PR DESCRIPTION
**Pull request by shay23b**

### CODE REVIEW

File: connect.go [hunk 1]
```diff
@@ -68,6 +70,37 @@ type ConfigurationsUpdate struct {
 	Update      bool   `json:"update"`
 }
 
+// FetchOpts - configuration options for fetch.
+type FetchOpts struct {
+	Name                     string
```
Comment (idanasulinmemphis) [inline connect.go:75]: change to consumerName

File: connect.go [hunk 2]
```diff
@@ -68,6 +70,37 @@ type ConfigurationsUpdate struct {
...
+	GenUniqueSuffix          bool
+	ErrHandler               ConsumerErrHandler
+	StartConsumeFromSequence uint64
+	LastMessages             int64
+}
+
+// getDefaultConsumerOptions - returns default configuration options for consumers.
+func getDefaultFetchOptions() FetchOpts {
+	return FetchOpts{
+		BatchMaxTimeToWait:       5 * time.Second,
```
Comment (idanasulinmemphis) [inline connect.go:91]: consumer group is also not mandatory default is "" and batch size is also can default to 10.
the batch size should only be mandatory on the fetch (of the consumer)

File: connect.go [hunk 3]
```diff
@@ -547,3 +591,145 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
 }
+func (cm *ConsumersMap) getConsumer(key string) *Consumer {
+	if (*cm) != nil && (*cm)[key] != nil {
+		return (*cm)[key]
+	}
+	return nil
+}
+
+func (cm *ConsumersMap) setConsumer(c *Consumer) {
+	cn := fmt.Sprintf("%s_%s", c.stationName, c.realName)
```
Comment (idanasulinmemphis) [inline connect.go:602]: please make sure the stationName is the internal one

File: connect.go [hunk 4]
```diff
@@ -547,3 +591,145 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
+	(*cm)[cn] = c
+}
+
+func (cm *ConsumersMap) unsetConsumer(key string) {
+	delete(*cm, key)
+}
+
+func (cm *ConsumersMap) unsetStationConsumers(stationName string) {
+	for k, v := range *cm {
+		if v.stationName == stationName {
```
Comment (idanasulinmemphis) [inline connect.go:615]: please make sure you are comparing same names (should be internal name)

File: connect.go [hunk 5]
```diff
@@ -547,3 +591,145 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
+			cm.unsetConsumer(k)
+		}
+	}
+}
+
+// FetchMessages - Consume a batch of messages.
+func (c *Conn) FetchMessages(stationName string, consumerName string, batchSize int, opts ...FetchOpt) ([]*Msg, error) {
+	var consumer *Consumer
+	cm := c.getConsumersMap()
+	cons := cm.getConsumer(fmt.Sprintf("%s_%s", strings.ToLower(stationName), strings.ToLower(consumerName)))
```
Comment (idanasulinmemphis) [inline connect.go:625]: station name should be the internal representation

File: connect.go [hunk 6]
```diff
@@ -547,3 +591,145 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
+	}
+	msgs, err := consumer.Fetch(batchSize)
+	if err != nil {
+		return nil, err
+	}
+	return msgs, nil
+}
+
+// ConsumerName - name for the consumer.
+func FetchConsumerName(name string) FetchOpt {
```
Comment (idanasulinmemphis) [inline connect.go:663]: this is not needed since it is a mandatory field

File: connect.go [hunk 7]
```diff
@@ -547,3 +591,145 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
+	return func(opts *FetchOpts) error {
+		opts.Name = name
+		return nil
+	}
+}
+
+// StationNameOpt - station name to consume messages from.
+func FetchStationNameOpt(stationName string) FetchOpt {
```
Comment (idanasulinmemphis) [inline connect.go:671]: this is not needed since it is a mandatory field

File: consumer.go [hunk 8]
```diff
@@ -415,36 +413,59 @@ func (c *Consumer) fetchSubscriprionWithTimeout() ([]*Msg, error) {
 	}()
 	select {
 	case <-time.After(timeoutDuration):
-		return nil, memphisError(errors.New("fetch timed out"))
+		return nil, nil
```
Comment (idanasulinmemphis) [inline consumer.go:416]: why nil in here? it seems like a timeout

File: consumer.go [hunk 9]
```diff
@@ -415,36 +413,59 @@ func (c *Consumer) fetchSubscriprionWithTimeout() ([]*Msg, error) {
...
 	case fetchRes := <-out:
 		return fetchRes.msgs, memphisError(fetchRes.err)
 
 	}
 }
 
 // Fetch - immediately fetch a message batch.
```
Comment (idanasulinmemphis) [inline consumer.go:423]: typo - messages

File: consumer.go [hunk 10]
```diff
@@ -198,14 +203,21 @@ func (c *Conn) CreateConsumer(stationName, consumerName string, opts ...Consumer
 			}
 		}
 	}
+	consumer, err := defaultOpts.createConsumer(c)
+	if err != nil {
+		return nil, memphisError(err)
+	}
+	cm := c.getConsumersMap()
+	cm.setConsumer(consumer)
```
Comment (idanasulinmemphis) [inline consumer.go:211]: why not calling cacheConsumer?

File: consumer.go (hunk 8 above)
Comment (shay23b) [reply consumer.go:416]: you are right, forgot it

File: README.md [hunk 11]
```diff
@@ -285,10 +285,22 @@ func handler(msgs []*memphis.Msg, err error, ctx context.Context) {
 consumer.Consume(handler)
 ```
 
-You can trigger a single fetch with the Fetch() method
+### Fetch a single batch of messages
+```go
+msgs, err := conn.FetchMessages("<station-name>", "<consumer-name>", <batch-size> int,
```
Comment (idanasulinmemphis) [inline README.md:290]: we turned it into optional

File: connect.go [hunk 12]
```diff
@@ -40,6 +40,8 @@ type Option func(*Options) error
 
 type ProducersMap map[string]*Producer
```
Comment (idanasulinmemphis) [inline connect.go:42]: remove the line space

File: connect.go [hunk 13]
```diff
@@ -547,3 +593,147 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
+	}
+	msgs, err := consumer.Fetch(defaultOpts.BatchSize)
+	if err != nil {
+		return nil, err
+	}
+	return msgs, nil
+}
+
+// ConsumerName - name for the consumer.
+func FetchConsumerName(name string) FetchOpt {
```
Comment (idanasulinmemphis) [inline connect.go:667]: I've commented that we don't need it since it is a mandatory field why have you left it?

File: connect.go [hunk 14]
```diff
@@ -547,3 +593,147 @@ func (pm *ProducersMap) unsetStationProducers(stationName string) {
...
+	return func(opts *FetchOpts) error {
+		opts.ConsumerName = name
+		return nil
+	}
+}
+
+// StationNameOpt - station name to consume messages from.
+func FetchStationNameOpt(stationName string) FetchOpt {
```
Comment (idanasulinmemphis) [inline connect.go:675]: same here

**Review by idanasulinmemphis — APPROVED** [review]

**Pull request merged**